ratik/src/intTest/java/org/opencadc/ratik/InventoryValidatorTest.java

```diff
@@ -0,0 +1,560 @@
+        Assert.assertNull("local artifact found", localArtifact);
+
+        DeletedStorageLocationEvent dsle = this.localEnvironment.deletedStorageLocationEventDAO.get(artifact.getID());
+        Assert.assertNotNull("DeletedStorageLocationEvent not found", dsle);
+    }
+
+    /** discrepancy: artifact in L && artifact not in R
+     * explanation1: deleted from R, pending/missed DeletedArtifactEvent in L
+     * evidence: DeletedArtifactEvent in R
+     * action: put DAE, delete artifact
+     *
+     * before: Artifact in L, not in R, DeletedArtifactEvent in R
+     * after: Artifact not in L, DeletedArtifactEvent in L
+     */
+    @Test
+    public void explanation1_inL() throws Exception {
+        Calendar calendar = Calendar.getInstance();
+        Artifact artifact = new Artifact(URI.create("cadc:INTTEST/one.ext"), TestUtil.getRandomMD5(), calendar.getTime(), 1024L);
+        this.localEnvironment.artifactDAO.put(artifact);
+
+        DeletedArtifactEvent deletedArtifactEvent = new DeletedArtifactEvent(artifact.getID());
+        this.remoteEnvironment.deletedArtifactEventDAO.put(deletedArtifactEvent);
+
+        InventoryValidator testSubject = new InventoryValidator(this.localEnvironment.daoConfig, TestUtil.LUSKAN_URI, artifact.getBucket());
+        testSubject.run();
+
+        Artifact localArtifact = this.localEnvironment.artifactDAO.get(artifact.getURI());
+        Assert.assertNull("local artifact found", localArtifact);
+
+        DeletedStorageLocationEvent dsle = this.localEnvironment.deletedStorageLocationEventDAO.get(artifact.getID());
+        Assert.assertNotNull("DeletedStorageLocationEvent not found", dsle);
+    }
+
+    /** discrepancy: artifact in L && artifact not in R
+     * explanation2: L==global, deleted from R, pending/missed DeletedStorageLocationEvent in L
+     * evidence: DeletedStorageLocationEvent in R
+     * action: remove siteID from Artifact.storageLocations (see note)
+     *
+     * note: when removing siteID from Artifact.storageLocations in global, if the Artifact.siteLocations becomes empty
+     * the artifact should be deleted (metadata-sync needs to also do this in response to a DeletedStorageLocationEvent)
+     * TBD: must this also create a DeletedArtifactEvent?
+     *
+     * L == global
+     * case 1
+     * before: Artifact in L with R siteID, plus others, in Artifact.siteLocations, not in R, DeletedStorageLocationEvent in R
+     * after: Artifact in L, R siteID not in Artifact.siteLocations
+     *
+     * case 2
+     * before: Artifact in L with R siteID in Artifact.siteLocations, not in R, DeletedStorageLocationEvent in R
+     * after: Artifact in not in L
+     */
+    @Test
+    public void explanation2_inL() throws Exception {
+        // case 1
+        UUID remoteSiteID = UUID.randomUUID();
+        UUID otherSiteID = UUID.randomUUID();
+
+        Calendar calendar = Calendar.getInstance();
```

Comment:
What's wrong with `new Date()`?


The SiteLocation is a reference to a StorageSite so the UUID is the StorageSite entity in the remoteEnvironment. That is normally put there by minoc (StorageSite is currently implicitly the resourceID of a "files" service) so it will not be there in these tests.

All the tests here are going to need to init the remoteEnvironment with a StorageSite when remote is storage because InventoryValidator is going to query luskan to get and use it (when L==global) just like fenwick does. See trackSiteLocations comment earlier.

Note: stopping review at this point because fixing this will likely cause other tests below to change but setup of the remote StorageSite may be needed above as well so I'll likely have to re-examine them.

Reply:
I thought there might be a need to change the lastModifiedDate to return artifacts in a different order, so used a Calendar, but not needed so far. Changed to Date.

Added a StorageSite to the remote site before each test.